In [244]:
from sklearn.linear_model import LogisticRegression
import numpy as np
import pandas as pd
import matplotlib

# Thu nhập dữ liệu

In [245]:
brcc_df = pd.read_csv('data.csv')
brcc_df

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0      842302         M        17.99         10.38          122.80     1001.0   
1      842517         M        20.57         17.77          132.90     1326.0   
2    84300903         M        19.69         21.25          130.00     1203.0   
3    84348301         M        11.42         20.38           77.58      386.1   
4    84358402         M        20.29         14.34          135.10     1297.0   
..        ...       ...          ...           ...             ...        ...   
564    926424         M        21.56         22.39          142.00     1479.0   
565    926682         M        20.13         28.25          131.20     1261.0   
566    926954         M        16.60         28.08          108.30      858.1   
567    927241         M        20.60         29.33          140.10     1265.0   
568     92751         B         7.76         24.54           47.92      181.0   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0            0.11840           0.27760         0.30010              0.14710   
1            0.08474           0.07864         0.08690              0.07017   
2            0.10960           0.15990         0.19740              0.12790   
3            0.14250           0.28390         0.24140              0.10520   
4            0.10030           0.13280         0.19800              0.10430   
..               ...               ...             ...                  ...   
564          0.11100           0.11590         0.24390              0.13890   
565          0.09780           0.10340         0.14400              0.09791   
566          0.08455           0.10230         0.09251              0.05302   
567          0.11780           0.27700         0.35140              0.15200   
568          0.05263           0.04362         0.00000              0.00000   

     ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0    ...          17.33           184.60      2019.0           0.16220   
1    ...          23.41           158.80      1956.0           0.12380   
2    ...          25.53           152.50      1709.0           0.14440   
3    ...          26.50            98.87       567.7           0.20980   
4    ...          16.67           152.20      1575.0           0.13740   
..   ...            ...              ...         ...               ...   
564  ...          26.40           166.10      2027.0           0.14100   
565  ...          38.25           155.00      1731.0           0.11660   
566  ...          34.12           126.70      1124.0           0.11390   
567  ...          39.42           184.60      1821.0           0.16500   
568  ...          30.37            59.16       268.6           0.08996   

     compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0              0.66560           0.7119                0.2654          0.4601   
1              0.18660           0.2416                0.1860          0.2750   
2              0.42450           0.4504                0.2430          0.3613   
3              0.86630           0.6869                0.2575          0.6638   
4              0.20500           0.4000                0.1625          0.2364   
..                 ...              ...                   ...             ...   
564            0.21130           0.4107                0.2216          0.2060   
565            0.19220           0.3215                0.1628          0.2572   
566            0.30940           0.3403                0.1418          0.2218   
567            0.86810           0.9387                0.2650          0.4087   
568            0.06444           0.0000                0.0000          0.2871   

     fractal_dimension_worst  Unnamed: 32  
0                    0.11890          NaN  
1                    0.08902          NaN  
2                    0.08758          NaN  
3                    0.17300          NaN  
4      

### Có bao nhiêu dòng cột? Có dòng nào bị lặp không ?

In [246]:
num_rows = len(brcc_df)
num_cols = len(brcc_df.columns)
print(num_rows , num_cols)

569 33


In [247]:
have_duplicated_rows = brcc_df.duplicated().any()
have_duplicated_rows

False

---

## Khám phá dữ liệu

In [248]:
#with open('description.txt', 'r',encoding='utf-8') as f:
    #print(f.read())

In [249]:
brcc_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 33 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       569 non-null    int64  
 1   diagnosis                569 non-null    object 
 2   radius_mean              569 non-null    float64
 3   texture_mean             569 non-null    float64
 4   perimeter_mean           569 non-null    float64
 5   area_mean                569 non-null    float64
 6   smoothness_mean          569 non-null    float64
 7   compactness_mean         569 non-null    float64
 8   concavity_mean           569 non-null    float64
 9   concave points_mean      569 non-null    float64
 10  symmetry_mean            569 non-null    float64
 11  fractal_dimension_mean   569 non-null    float64
 12  radius_se                569 non-null    float64
 13  texture_se               569 non-null    float64
 14  perimeter_se             5

Bởi vì có 569 dòng và 33 cột, các trường đều thể hiện đầy đủ dữ liệu trừ cột cuối cùng "Unamed: 32". 

Hình như cột dữ liệu này pandas lấy nhầm từ tập dữ liệu đã cho, ta sẽ handle cột dữ liệu này trong phần Tiền xử lý dữ liệu.

In [250]:
brcc_df.dtypes

id                           int64
diagnosis                   object
radius_mean                float64
texture_mean               float64
perimeter_mean             float64
area_mean                  float64
smoothness_mean            float64
compactness_mean           float64
concavity_mean             float64
concave points_mean        float64
symmetry_mean              float64
fractal_dimension_mean     float64
radius_se                  float64
texture_se                 float64
perimeter_se               float64
area_se                    float64
smoothness_se              float64
compactness_se             float64
concavity_se               float64
concave points_se          float64
symmetry_se                float64
fractal_dimension_se       float64
radius_worst               float64
texture_worst              float64
perimeter_worst            float64
area_worst                 float64
smoothness_worst           float64
compactness_worst          float64
concavity_worst     

Hầu hết các trường dữ liệu đều là float nên sẽ không có cột nào categorical, tất cả được xem là numerical ngoại trừ hai cột attribute của từng người là ID và Diagnosis.

In [251]:
numerical_cols = list(brcc_df.iloc[:,2:num_cols].columns)
atribute_cols = list(brcc_df.iloc[:,0:2].columns)

### Làm sạch dữ liệu (Data cleaning)

Bởi vì không có dòng trong trường dữ liệu nào là null. Nhưng điều này hơi kì lạ nếu dữ liệu không bị thiếu sót gì, ta cần thực hiện kiểm tra các trường xem tỷ lệ các giá trị bằng 0. Bởi vì đối với các dữ liệu sinh học về cơ thể con người, các giá trị cần được làm rõ.

In [252]:
for col in numerical_cols:
    s = brcc_df[col].value_counts()
    try:
        print(col,s[0])
    except:
        continue

concavity_mean 13
concave points_mean 13
concavity_se 13
concave points_se 13
concavity_worst 13
concave points_worst 13


Như ta có thể thấy, tất cả các trường liên quan đến concavity nếu có giá trị 0 thì các trường còn lại đều có 0. Tiếp theo, ta phải check xem nếu một cột của trường này bằng 0 thì tất cả các cột còn lại có bằng 0 không?

In [253]:
zero_cols = ['concavity_mean','concave points_mean','concavity_se','concave points_se','concavity_worst','concave points_worst']
test_0 = brcc_df[zero_cols]

In [254]:
test_0[test_0['concavity_mean'] == 0]

concavity_mean  concave points_mean  concavity_se  concave points_se  \
101             0.0                  0.0           0.0                0.0   
140             0.0                  0.0           0.0                0.0   
174             0.0                  0.0           0.0                0.0   
175             0.0                  0.0           0.0                0.0   
192             0.0                  0.0           0.0                0.0   
314             0.0                  0.0           0.0                0.0   
391             0.0                  0.0           0.0                0.0   
473             0.0                  0.0           0.0                0.0   
538             0.0                  0.0           0.0                0.0   
550             0.0                  0.0           0.0                0.0   
557             0.0                  0.0           0.0                0.0   
561             0.0                  0.0           0.0                0.0   
568             0.0                  0.0           0.0                0.0   

     concavity_worst  concave points_worst  
101              0.0                   0.0  
140              0.0                   0.0  
174              0.0                   0.0  
175              0.0                   0.0  
192              0.0                   0.0  
314              0.0                   0.0  
391              0.0                   0.0  
473              0.0                   0.0  
538              0.0                   0.0  
550              0.0                   0.0  
557              0.0                   0.0  
561              0.0                   0.0  
568              0.0                   0.0

In [255]:
test_0[test_0['concavity_worst'] == 0]

concavity_mean  concave points_mean  concavity_se  concave points_se  \
101             0.0                  0.0           0.0                0.0   
140             0.0                  0.0           0.0                0.0   
174             0.0                  0.0           0.0                0.0   
175             0.0                  0.0           0.0                0.0   
192             0.0                  0.0           0.0                0.0   
314             0.0                  0.0           0.0                0.0   
391             0.0                  0.0           0.0                0.0   
473             0.0                  0.0           0.0                0.0   
538             0.0                  0.0           0.0                0.0   
550             0.0                  0.0           0.0                0.0   
557             0.0                  0.0           0.0                0.0   
561             0.0                  0.0           0.0                0.0   
568             0.0                  0.0           0.0                0.0   

     concavity_worst  concave points_worst  
101              0.0                   0.0  
140              0.0                   0.0  
174              0.0                   0.0  
175              0.0                   0.0  
192              0.0                   0.0  
314              0.0                   0.0  
391              0.0                   0.0  
473              0.0                   0.0  
538              0.0                   0.0  
550              0.0                   0.0  
557              0.0                   0.0  
561              0.0                   0.0  
568              0.0                   0.0

Nếu một dòng trong một trường dữ liệu bất kì trong zero_cols đã khai báo bằng 0, tất cả các dòng còn lại ở các trường dữ liệu còn lại đều bằng 0.

Vậy ta sẽ thực hiện làm sạch dữ liệu bằng cách thay thế các giá trị bằng 0 này hoặc bỏ luôn toàn bộ các dòng này.

In [256]:
round(len(brcc_df[brcc_df['concavity_worst'] == 0]) / len(brcc_df) * 100, 2)

2.28

Bởi vì tỷ lệ của các dòng đã nêu chỉ chiếm hơn 2% nên quyết định drop luôn là lựa chọn hợp lý.

In [257]:
brcc_df = brcc_df[brcc_df['concavity_worst'] != 0]

### Tiền xử lý dữ liệu

Tiền xử lý thì không có gì nhiều để làm, các dữ liệu đều là số và cách biệt giữa các giá trị không quá nhiều nên ta không cần chuẩn hóa.

Tuy nhiên, khi áp dụng mô hình hồi quy cụ thể là Logistic Regression, ta cần các dữ liệu số và không cần quan tâm đến các giá trị attribute của từng người, cột ID ta có thể drop. Còn cột Diagnosis, ta đổi các giá trị bên trong thành (0,1) để áp dụng các hàm hồi quy khi phân hóa dữ liệu.

Cột Diagnosis có hai giá trị chính là:
- M: malignant(nặng) --> M = 1
- B: bengin(nhẹ)     --> B = 0

Ngoài ra, đã nhắc ở phần Khám phá dữ liệu, ta sẽ drop luôn cột cuối cùng vì có thể cột dữ liệu này pandas lấy nhầm từ tập dữ liệu đã cho.

In [258]:
brcc_df = brcc_df.drop(columns=['id','Unnamed: 32'],axis=1)

In [259]:
brcc_df[atribute_cols[1]] = brcc_df[atribute_cols[1]].replace(to_replace=[['B'],['M']],value=[0,1])

In [260]:
brcc_df

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0            1        17.99         10.38          122.80     1001.0   
1            1        20.57         17.77          132.90     1326.0   
2            1        19.69         21.25          130.00     1203.0   
3            1        11.42         20.38           77.58      386.1   
4            1        20.29         14.34          135.10     1297.0   
..         ...          ...           ...             ...        ...   
563          1        20.92         25.09          143.00     1347.0   
564          1        21.56         22.39          142.00     1479.0   
565          1        20.13         28.25          131.20     1261.0   
566          1        16.60         28.08          108.30      858.1   
567          1        20.60         29.33          140.10     1265.0   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0            0.11840           0.27760         0.30010              0.14710   
1            0.08474           0.07864         0.08690              0.07017   
2            0.10960           0.15990         0.19740              0.12790   
3            0.14250           0.28390         0.24140              0.10520   
4            0.10030           0.13280         0.19800              0.10430   
..               ...               ...             ...                  ...   
563          0.10990           0.22360         0.31740              0.14740   
564          0.11100           0.11590         0.24390              0.13890   
565          0.09780           0.10340         0.14400              0.09791   
566          0.08455           0.10230         0.09251              0.05302   
567          0.11780           0.27700         0.35140              0.15200   

     symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0           0.2419  ...         25.38          17.33           184.60   
1           0.1812  ...         24.99          23.41           158.80   
2           0.2069  ...         23.57          25.53           152.50   
3           0.2597  ...         14.91          26.50            98.87   
4           0.1809  ...         22.54          16.67           152.20   
..             ...  ...           ...            ...              ...   
563         0.2149  ...         24.29          29.41           179.10   
564         0.1726  ...         25.45          26.40           166.10   
565         0.1752  ...         23.69          38.25           155.00   
566         0.1590  ...         18.98          34.12           126.70   
567         0.2397  ...         25.74          39.42           184.60   

     area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0        2019.0            0.1622             0.6656           0.7119   
1        1956.0            0.1238             0.1866           0.2416   
2        1709.0            0.1444             0.4245           0.4504   
3         567.7            0.2098             0.8663           0.6869   
4        1575.0            0.1374             0.2050           0.4000   
..          ...               ...                ...              ...   
563      1819.0            0.1407             0.4186           0.6599   
564      2027.0            0.1410             0.2113           0.4107   
565      1731.0            0.1166             0.1922           0.3215   
566      1124.0            0.1139             0.3094           0.3403   
567      1821.0            0.1650             0.8681           0.9387   

     concave points_worst  symmetry_worst  fractal_dimension_worst  
0                  0.2654          0.4601                  0.11890  
1                  0.1860          0.2750                  0.08902  
2                  0.2430          0.3613                  0.08758  
3                  0.2575          0.6638                  0.17300  
4                  0.1625          0.2364                  0.07678  
..                    .

Tiền xử lý dữ liệu đã xong.

### Phân tích dữ liệu (Trần Quốc Việt)

### Xây dựng mô hình và dự đoán